### PR TITLE
[Feat][CONV] Consolidate conv kernels and benchmark defaults

### LINKED ISSUE #667: [FEAT][CONV] implement conv1d, conv3d
### Symptom / Motivation
Issue #402 tracks the implementation of the conv & pooling operator family. `conv2d` has already landed in TileOPs, but the standard convolution family is still incomplete because `conv1d` and `conv3d` are missing.

These two operators are still important coverage gaps:
- `conv1d` is required for audio, time-series, and sequence-style convolution workloads
- `conv3d` is required for video, volumetric, and spatiotemporal convolution workloads

Completing them brings the standard convolution set in TileOPs to parity with the three core PyTorch convolution APIs.

### Root Cause Analysis
N/A - Feature implementation (new operators).

The repository already contains the `conv2d` kernel/op/test/benchmark path, but there are no corresponding `conv1d` or `conv3d` implementations yet.

### Related Files
- `tileops/kernels/conv2d/conv2d.py` — existing reference implementation pattern for standard convolution kernel
- `tileops/kernels/conv1d/` — add: kernel package for Conv1d
- `tileops/kernels/conv3d/` — add: kernel package for Conv3d
- `tileops/ops/conv2d.py` — existing reference implementation pattern for op wrapper
- `tileops/ops/conv1d.py` — add: Op wrapper for Conv1d
- `tileops/ops/conv3d.py` — add: Op wrapper for Conv3d
- `tileops/kernels/__init__.py` — modify: export new kernel classes or helpers if needed
- `tileops/ops/__init__.py` — modify: export new ops
- `tests/ops/test_conv1d.py` — add: correctness tests
- `tests/ops/test_conv3d.py` — add: correctness tests
- `benchmarks/ops/bench_conv1d.py` — add: performance benchmarks
- `benchmarks/ops/bench_conv3d.py` — add: performance benchmarks

## Goal
Implement `conv1d` and `conv3d` following the existing TileOPs 2-layer architecture (Kernel L1 -> Op L2), using `conv2d` as the primary in-repo implementation reference and validating correctness against PyTorch.

## Plan

1. Study the existing `conv2d` kernel and op implementation to extract reusable shape, launch, and API patterns.
2. Implement TileLang kernels for `conv1d` and `conv3d` with standard convolution semantics aligned with `torch.nn.functional.conv1d` and `torch.nn.functional.conv3d`.
3. Add L2 op wrappers for both operators with the same integration style as existing TileOPs ops.
4. Update exports in `tileops/kernels/__init__.py` and `tileops/ops/__init__.py` as needed.
5. Add correctness tests covering representative shapes, stride/padding settings, and FP16/BF16 comparison against PyTorch reference.
6. Add benchmarks reporting latency, TFLOPS, and DRAM bandwidth for representative 1D and 3D convolution workloads.

## Constraints
- Must follow TileOPs 2-layer architecture (Kernel L1 -> Op L2)
- Must follow existing `conv2d` structure and naming conventions where applicable
- CUDA only; CPU support is out of scope
- Cannot change any existing public APIs or break existing tests
- Parent issue: #402

## Acceptance Criteria
- [ ] Modified files pass unit tests
- [ ] AC-1: Conv1d forward kernel and Op are implemented and pass PyTorch reference tests for FP16 and BF16
- [ ] AC-2: Conv3d forward kernel and Op are implemented and pass PyTorch reference tests for FP16 and BF16
- [ ] AC-3: `conv1d` and `conv3d` are exported through the standard TileOPs op surface
- [ ] AC-4: Both ops include benchmarks reporting latency, TFLOPS, and DRAM bandwidth
- [ ] AC-5: The standard convolution family tracked by #402 is complete for `conv1d`, `conv2d`, and `conv3d`

### PR DESCRIPTION
**Pull request by RMLYC**
Closes #667

## Summary

- consolidate `conv1d`, `conv2d`, and `conv3d` kernels under `tileops/kernels/conv` with a single export surface
- extract shared convolution autotune memory sizing into `tileops/kernels/conv/common.py`
- restrict conv autotune search space to `enable_rasteration=True`
- add comments on kernel-side `weight.permute(...)` layout transforms
- remove explicit `warmup=5, rep=10` from conv benchmarks so they use `BenchmarkBase.profile` defaults

## Test plan

- [x] `python -m py_compile benchmarks/ops/bench_conv1d.py benchmarks/ops/bench_conv2d.py benchmarks/ops/bench_conv3d.py`
- [x] `CUDA_VISIBLE_DEVICES=1 python -m pytest tests/ops/test_conv1d.py tests/ops/test_conv2d.py tests/ops/test_conv3d.py -vvs`
- [x] `CUDA_VISIBLE_DEVICES=1 python -m pytest benchmarks/ops/bench_conv1d.py benchmarks/ops/bench_conv3d.py -vvs`

### conv1d tileops vs torch

| case | tileops tflops | torch tflops | speedup |
| --- | ---: | ---: | ---: |
| convtasnet-pointwise-k1-s1-fp16 | 128.25 | 62.37 | 2.06x |
| seanet-k3-s1-fp16 | 139.42 | 46.72 | 2.98x |
| audio-downsample-k5-s2-fp16 | 111.18 | 30.52 | 3.64x |
| seanet-stem-k7-s1-fp16 | 189.80 | 97.15 | 1.95x |
| sequence-downsample-k3-s2-bf16 | 66.38 | 24.80 | 2.68x |

### conv3d tileops vs torch

| case | tileops tflops | torch-ncdhw tflops | torch-ndhwc tflops | vs best torch |
| --- | ---: | ---: | ---: | ---: |
| r3d-stem-k3-s1-fp16 | 24.40 | 8.88 | 10.88 | 2.24x |
| video-stage-downsample-k3-s2-fp16 | 54.48 | 5.43 | 44.78 | 1.22x |
| unet-encoder-k3-s1-bf16 | 112.96 | 33.13 | 70.13 | 1.61x |

## Regression

- `29 passed` for conv unit tests
- `8 passed` for `bench_conv1d.py` and `bench_conv3d.py`

### CODE REVIEW

File: benchmarks/ops/bench_conv3d.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            x,
+            weight,
+            bias=bias,
+            stride=self.stride,
+            padding=self.padding,
+            dilation=1,
+            groups=1,
+        )
+
+    def ref_program_ndhwc(
```
Comment (superAngGao) [inline benchmarks/ops/bench_conv3d.py]: `ref_program_ndhwc` and `ref_program_ncdhw` (line 49) have identical implementations. The performance difference comes from the caller passing tensors with different `memory_format`, not from anything in the method body.

Consider either:
- Merging into a single `ref_program` (like `bench_conv1d.py` does), or
- Adding a brief comment explaining the bodies are intentionally identical and that the layout difference is in the caller

File: tileops/kernels/conv/conv1d.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        weight: torch.Tensor,
+        bias: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        if bias is None:
+            bias = torch.zeros(self.c_out, device=x.device, dtype=x.dtype)
+        # OCI -> KIO so the kernel can flatten weights into [K_total, C_out].
```
Comment (superAngGao) [inline tileops/kernels/conv/conv1d.py]: Nit: The comment says `OCI -> KIO`, but the weight shape is `(c_out, c_in, kernel_l)` and `permute(2, 1, 0)` yields `(kernel_l, c_in, c_out)`. In standard notation this is `OIK -> KIO` (O=c_out, I=c_in, K=kernel_l).

The `C` in `OCI` is ambiguous — it could mean either channels or something else.

File: tileops/kernels/conv/conv3d.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        weight: torch.Tensor,
+        bias: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        if bias is None:
+            bias = torch.zeros(self.c_out, device=x.device, dtype=x.dtype)
+        # OIDHW -> KDHWIO so the kernel can flatten weights into [K_total, C_out].
```
Comment (superAngGao) [inline tileops/kernels/conv/conv3d.py]: The comment says `OIDHW -> KDHWIO`, but `KDHWIO` has 6 characters for a 5-dim tensor. `permute(2, 3, 4, 1, 0)` maps `(O, I, D, H, W) -> (D, H, W, I, O)`, so it should be `OIDHW -> DHWIO`.

File: tileops/kernels/conv/conv1d.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    def default_config(self) -> dict:
+        sm_version = get_sm_version()
+        if sm_version in {90}:
+            return {
+                "block_m": 128,
+                "block_n": 64,
+                "block_k": 64,
+                "num_stages": 3,
+                "threads": 128,
+                "enable_rasteration": False,
```
Comment (superAngGao) [inline tileops/kernels/conv/conv1d.py]: On sm90, `default_config` uses `enable_rasteration: False`, but `autotune_configs` only generates `[True]`. This means the default config can never be selected by autotuning — is that intentional?

Same pattern in `Conv3dKernel`. If `True` is consistently better (which the autotune-only-True restriction suggests), consider using it in the default as well.

File: tileops/kernels/conv/common.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import torch
+
+__all__ = ["HOPPER_SHARED_MEMORY_LIMIT_BYTES", "conv_shared_memory_bytes"]
+
+HOPPER_SHARED_MEMORY_LIMIT_BYTES = 227 * 1024
```
Comment (superAngGao) [inline tileops/kernels/conv/common.py]: This 227KB limit is Hopper-specific, but `supported_archs` in Conv1dKernel/Conv3dKernel includes sm80/86/89 which have lower shared memory limits (e.g. Ampere sm80 = 164KB, sm86 = 100KB). Autotune configs that pass this filter may still exceed shared memory on non-Hopper GPUs.

Not a blocker (this was inherited from conv2d), but now that this is a shared constant, it might be worth making it arch-aware — e.g. a `get_shared_memory_limit()` helper that checks `get_sm_version()`.

File: benchmarks/ops/bench_conv3d.py (hunk 1 above)
Comment (RMLYC) [reply benchmarks/ops/bench_conv3d.py]: Addressed in `86d6447`. I merged the duplicated `ref_program_ncdhw` / `ref_program_ndhwc` helpers into a single `ref_program`, since the layout difference is in the caller-provided tensor memory format rather than in the method body.

File: tileops/kernels/conv/conv1d.py (hunk 2 above)
Comment (RMLYC) [reply tileops/kernels/conv/conv1d.py]: Addressed in `86d6447`. Updated the comment to `OIK -> KIO` so the dimension labels match the actual `(c_out, c_in, kernel_l) -> (kernel_l, c_in, c_out)` permutation.

File: tileops/kernels/conv/conv3d.py (hunk 3 above)
Comment (RMLYC) [reply tileops/kernels/conv/conv3d.py]: Addressed in `86d6447`. Updated the comment to `OIDHW -> DHWIO` to match the `permute(2, 3, 4, 1, 0)` layout transform.

File: tileops/kernels/conv/conv1d.py (hunk 4 above)
Comment (RMLYC) [reply tileops/kernels/conv/conv1d.py]: Addressed in `86d6447`. I aligned the default configs with the autotune search space by switching the SM90 `enable_rasteration` default to `True` in both `Conv1dKernel` and `Conv3dKernel`.

File: tileops/kernels/conv/common.py (hunk 5 above)
Comment (RMLYC) [reply tileops/kernels/conv/common.py]: Addressed in `86d6447`. I replaced the Hopper-only shared-memory constant with a runtime helper that reads `torch.cuda.get_device_properties(...).shared_memory_per_block_optin`, so the autotune filter now uses the actual device limit. I also wired the same helper through the shared conv path, which fixes the inherited behavior for conv2d as well.
